Давайте проанализируем данные опроса 4361 женщин из Ботсваны:

О каждой из них мы знаем:

* сколько детей она родила (признак `ceb`)
* возраст (`age`)
* длительность получения образования (`educ`)
* религиозная принадлежность (`religion`)
* идеальное, по её мнению, количество детей в семье (`idlnchld`)
* была ли она когда-нибудь замужем (`evermarr`)
* возраст первого замужества (`agefm`)
* длительность получения образования мужем (`heduc`)
* знает ли она о методах контрацепции (`knowmeth`)
* использует ли она методы контрацепции (`usemeth`)
* живёт ли она в городе (`urban`)
* есть ли у неё электричество, радио, телевизор и велосипед (`electric`, `radio`, `tv`, `bicycle`)

Давайте научимся оценивать количество детей `ceb` по остальным признакам.

Загрузите данные и внимательно изучите их. Сколько разных значений принимает признак `religion`?

In [62]:
import pandas

frame = pandas.read_table('botswana.tsv', sep='\t')
frame.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  evermarr  agefm  \
0    0   18    10    catholic       4.0       1.0      1.0         0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0         1   20.0   
2    0   49     4      spirit       4.0       1.0      0.0         1   22.0   
3    0   24    12       other       2.0       1.0      0.0         0    NaN   
4    3   32    13       other       3.0       1.0      1.0         1   24.0   

   heduc  urban  electric  radio   tv  bicycle  
0    NaN      1       1.0    1.0  1.0      1.0  
1   14.0      1       1.0    1.0  1.0      1.0  
2    1.0      1       1.0    1.0  0.0      0.0  
3    NaN      1       1.0    1.0  1.0      1.0  
4   12.0      1       1.0    1.0  1.0      1.0

In [63]:
pandas.DataFrame(frame['religion'].value_counts())

religion
spirit          1841
other           1080
protestant       993
catholic         447

Во многих признаках есть пропущенные значения. Сколько объектов из 4361 останется, если выбросить все, содержащие пропуски?

In [64]:
frame.dropna().shape

(1834, 15)

* Создайте признак `nevermarr`, равный единице там, где в `agefm` пропуски.
* Удалите признак `evermarr` — в сумме с nevermarr он даёт константу, значит, в нашей матрице `X` будет мультиколлинеарность.
* Замените `NaN` в признаке `agefm` на `0`.
* У объектов, где `nevermarr = 1`, замените `NaN` в признаке `heduc` на `−1` (ноль использовать нельзя, так как он уже встречается у некоторых объектов выборки).

Сколько осталось пропущенных значений в признаке `heduc`?

In [65]:
frame['nevermarr'] = frame['agefm'].isnull().astype(int)
frame.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  evermarr  agefm  \
0    0   18    10    catholic       4.0       1.0      1.0         0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0         1   20.0   
2    0   49     4      spirit       4.0       1.0      0.0         1   22.0   
3    0   24    12       other       2.0       1.0      0.0         0    NaN   
4    3   32    13       other       3.0       1.0      1.0         1   24.0   

   heduc  urban  electric  radio   tv  bicycle  nevermarr  
0    NaN      1       1.0    1.0  1.0      1.0          1  
1   14.0      1       1.0    1.0  1.0      1.0          0  
2    1.0      1       1.0    1.0  0.0      0.0          0  
3    NaN      1       1.0    1.0  1.0      1.0          1  
4   12.0      1       1.0    1.0  1.0      1.0          0

In [66]:
frame.drop(['evermarr'], axis=1, inplace=True)
frame.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  agefm  heduc  \
0    0   18    10    catholic       4.0       1.0      1.0    NaN    NaN   
1    2   43    11  protestant       2.0       1.0      1.0   20.0   14.0   
2    0   49     4      spirit       4.0       1.0      0.0   22.0    1.0   
3    0   24    12       other       2.0       1.0      0.0    NaN    NaN   
4    3   32    13       other       3.0       1.0      1.0   24.0   12.0   

   urban  electric  radio   tv  bicycle  nevermarr  
0      1       1.0    1.0  1.0      1.0          1  
1      1       1.0    1.0  1.0      1.0          0  
2      1       1.0    1.0  0.0      0.0          0  
3      1       1.0    1.0  1.0      1.0          1  
4      1       1.0    1.0  1.0      1.0          0

In [67]:
frame['agefm'].fillna(0.0, inplace=True)
frame.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  agefm  heduc  \
0    0   18    10    catholic       4.0       1.0      1.0    0.0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0   20.0   14.0   
2    0   49     4      spirit       4.0       1.0      0.0   22.0    1.0   
3    0   24    12       other       2.0       1.0      0.0    0.0    NaN   
4    3   32    13       other       3.0       1.0      1.0   24.0   12.0   

   urban  electric  radio   tv  bicycle  nevermarr  
0      1       1.0    1.0  1.0      1.0          1  
1      1       1.0    1.0  1.0      1.0          0  
2      1       1.0    1.0  0.0      0.0          0  
3      1       1.0    1.0  1.0      1.0          1  
4      1       1.0    1.0  1.0      1.0          0

In [68]:
frame.loc[frame['nevermarr'] == 1, 'heduc'] = -1
frame.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  agefm  heduc  \
0    0   18    10    catholic       4.0       1.0      1.0    0.0   -1.0   
1    2   43    11  protestant       2.0       1.0      1.0   20.0   14.0   
2    0   49     4      spirit       4.0       1.0      0.0   22.0    1.0   
3    0   24    12       other       2.0       1.0      0.0    0.0   -1.0   
4    3   32    13       other       3.0       1.0      1.0   24.0   12.0   

   urban  electric  radio   tv  bicycle  nevermarr  
0      1       1.0    1.0  1.0      1.0          1  
1      1       1.0    1.0  1.0      1.0          0  
2      1       1.0    1.0  0.0      0.0          0  
3      1       1.0    1.0  1.0      1.0          1  
4      1       1.0    1.0  1.0      1.0          0

In [69]:
frame['heduc'].isnull().sum()

123

Избавимся от оставшихся пропусков.

Для признаков `idlnchld`, `heduc` и `usemeth` проведите операцию, аналогичную предыдущей: создайте индикаторы пропусков по этим признакам (`idlnchld_noans`, `heduc_noans`, `usemeth_noans`), замените пропуски на нехарактерные значения (`c_idlnchld = −1`, `c_heduc2 = −2` (значение `-1` мы уже использовали), `c_usemeth = −1`).

In [70]:
def fix_nan(column: str, new_column: str, value):
    frame[new_column] = frame[column].isnull().astype(int)
    frame.loc[frame[new_column] == 1, column] = value
    
fix_nan('idlnchld', 'idlnchld_noans', -1)    
fix_nan('heduc', 'heduc_noans', -2)  
fix_nan('usemeth', 'usemeth_noans', -1)  

Остались только пропуски в признаках `knowmeth`, `electric`, `radio`, `tv` и `bicycle`. Их очень мало, так что удалите объекты, на которых их значения пропущены.

In [71]:
frame.dropna(subset=['knowmeth', 'electric', 'radio', 'tv', 'bicycle'], inplace=True)

Какого размера теперь наша матрица данных? Умножьте количество строк на количество всех столбцов (включая отклик `ceb`).

In [72]:
rows, cols = frame.shape
rows * cols

78264

Постройте регрессию количества детей `ceb` на все имеющиеся признаки методом `smf.ols`, как в разобранном до этого примере. Какой получился коэффициент детерминации `R2`? Округлите до трёх знаков после десятичной точки.

In [73]:
' + '.join(frame.columns)

'ceb + age + educ + religion + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric + radio + tv + bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans'

In [74]:
from statsmodels.formula import api as formula_api

model = formula_api.ols('ceb ~ age + educ + religion + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric + radio + tv + bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans', data=frame)
fitted = model.fit()
fitted.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    ceb   R-squared:                       0.644
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     412.5
Date:                Sat, 24 Feb 2018   Prob (F-statistic):               0.00
Time:                        16:43:10   Log-Likelihood:                -7732.1
No. Observations:                4348   AIC:                         1.550e+04
Df Residuals:                    4328   BIC:                         1.563e+04
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -1.0263      0.212     -4.835      0.000      -1.443      -0.610
religion[T.other]         -0.0830      0.083     -1.001      0.317      -0.245       0.080
religion[T.protestant]    -0.0149      0.082     -0.181      0.857      -0.176       0.146
religion[T.spirit]        -0.0191      0.077     -0.248      0.804      -0.171       0.132
age                        0.1703      0.003     51.891      0.000       0.164       0.177
educ                      -0.0724      0.007     -9.843      0.000      -0.087      -0.058
idlnchld                   0.0760      0.011      6.923      0.000       0.054       0.098
knowmeth                   0.5564      0.121      4.580      0.000       0.318       0.795
usemeth                    0.6473      0.048     13.424      0.000       0.553       0.742
agefm                     -0.0604      0.007     -9.213      0.000      -0.073      -0.048
heduc                     -0.0551      0.008     -6.838      0.000      -0.071      -0.039
urban                     -0.2137      0.047     -4.527      0.000      -0.306      -0.121
electric                  -0.2685      0.077     -3.479      0.001      -0.420      -0.117
radio                     -0.0235      0.051     -0.461      0.645      -0.123       0.076
tv                        -0.1451      0.093     -1.566      0.118      -0.327       0.037
bicycle                    0.2139      0.050      4.260      0.000       0.115       0.312
nevermarr                 -2.2393      0.148    -15.143      0.000      -2.529      -1.949
idlnchld_noans             0.6539      0.153      4.286      0.000       0.355       0.953
heduc_noans               -0.8724      0.145     -6.026      0.000      -1.156      -0.589
usemeth_noans              0.7652      0.196      3.910      0.000       0.382       1.149
==============================================================================
Omnibus:                      224.411   Durbin-Watson:                   1.887
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              859.014
Skew:                           0.003   Prob(JB):                    2.93e-187
Kurtosis:                       5.178   Cond. No.                         361.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Проверьте критерием Бройша-Пагана гомоскедастичность ошибки в построенной модели. Выполняется ли она?

In [75]:
from statsmodels.stats import api as stats_api

_, p, *_ = stats_api.het_breuschpagan(fitted.resid, fitted.model.exog)
p

1.1452927633437192e-225

Если ошибка гетероскедастична, перенастройте модель, сделав поправку Уайта типа `HC1`.

In [76]:
model = formula_api.ols('ceb ~ age + educ + religion + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric + radio + tv + bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans', data=frame)
fitted = model.fit(cov_type='HC1')
fitted.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    ceb   R-squared:                       0.644
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     345.0
Date:                Sat, 24 Feb 2018   Prob (F-statistic):               0.00
Time:                        16:43:10   Log-Likelihood:                -7732.1
No. Observations:                4348   AIC:                         1.550e+04
Df Residuals:                    4328   BIC:                         1.563e+04
Df Model:                          19                                         
Covariance Type:                  HC1                                         
==========================================================================================
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -1.0263      0.266     -3.863      0.000      -1.547      -0.506
religion[T.other]         -0.0830      0.078     -1.067      0.286      -0.235       0.069
religion[T.protestant]    -0.0149      0.078     -0.192      0.848      -0.167       0.137
religion[T.spirit]        -0.0191      0.071     -0.268      0.789      -0.159       0.121
age                        0.1703      0.004     38.627      0.000       0.162       0.179
educ                      -0.0724      0.007     -9.924      0.000      -0.087      -0.058
idlnchld                   0.0760      0.015      5.236      0.000       0.048       0.104
knowmeth                   0.5564      0.174      3.190      0.001       0.215       0.898
usemeth                    0.6473      0.052     12.478      0.000       0.546       0.749
agefm                     -0.0604      0.010     -6.174      0.000      -0.080      -0.041
heduc                     -0.0551      0.009     -6.126      0.000      -0.073      -0.037
urban                     -0.2137      0.046     -4.667      0.000      -0.303      -0.124
electric                  -0.2685      0.072     -3.732      0.000      -0.410      -0.128
radio                     -0.0235      0.053     -0.446      0.656      -0.127       0.080
tv                        -0.1451      0.082     -1.766      0.077      -0.306       0.016
bicycle                    0.2139      0.048      4.412      0.000       0.119       0.309
nevermarr                 -2.2393      0.202    -11.082      0.000      -2.635      -1.843
idlnchld_noans             0.6539      0.216      3.029      0.002       0.231       1.077
heduc_noans               -0.8724      0.191     -4.556      0.000      -1.248      -0.497
usemeth_noans              0.7652      0.213      3.590      0.000       0.347       1.183
==============================================================================
Omnibus:                      224.411   Durbin-Watson:                   1.887
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              859.014
Skew:                           0.003   Prob(JB):                    2.93e-187
Kurtosis:                       5.178   Cond. No.                         361.
==============================================================================

Warnings:
[1] Standard Errors are heteroscedasticity robust (HC1)
"""

Удалите из модели незначимые признаки `religion`, `radio` и `tv`. Проверьте гомоскедастичность ошибки, при необходимости сделайте поправку Уайта.

In [77]:
new_model = formula_api.ols('ceb ~ age + educ + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric + bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans', data=frame)
_, p, _ = model.fit().compare_f_test(new_model.fit())
print(f'{p:.4f}')

0.4672


In [78]:
model = new_model

Удалите из текущей модели `usemeth_noans` и `usemeth`. Проверьте критерием Фишера гипотезу о том, что качество модели не ухудшилось. Введите номер первой значащей цифры в достигаемом уровне значимости.

In [80]:
new_model = formula_api.ols('ceb ~ age + educ + idlnchld + knowmeth + agefm + heduc + urban + electric + bicycle + nevermarr + idlnchld_noans + heduc_noans', data=frame)
_, p, _ = model.fit().compare_f_test(new_model.fit())
print(f'{p:E}')

3.155201E-40


Посмотрите на доверительные интервалы для коэффициентов итоговой модели (не забудьте использовать поправку Уайта, если есть гетероскедастичность ошибки) и выберите правильные выводы.

In [81]:
fitted = model.fit()
_, p, *_ = stats_api.het_breuschpagan(fitted.resid, fitted.model.exog)
p

1.1197458896536614e-228

In [86]:
model.fit(cov_type='HC1').summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    ceb   R-squared:                       0.644
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     463.4
Date:                Sat, 24 Feb 2018   Prob (F-statistic):               0.00
Time:                        17:18:27   Log-Likelihood:                -7734.5
No. Observations:                4348   AIC:                         1.550e+04
Df Residuals:                    4333   BIC:                         1.559e+04
Df Model:                          14                                         
Covariance Type:                  HC1                                         
==================================================================================
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -1.0698      0.258     -4.152      0.000      -1.575      -0.565
age                0.1702      0.004     38.746      0.000       0.162       0.179
educ              -0.0729      0.007    -10.311      0.000      -0.087      -0.059
idlnchld           0.0770      0.014      5.323      0.000       0.049       0.105
knowmeth           0.5610      0.174      3.224      0.001       0.220       0.902
usemeth            0.6516      0.052     12.571      0.000       0.550       0.753
agefm             -0.0606      0.010     -6.192      0.000      -0.080      -0.041
heduc             -0.0573      0.009     -6.440      0.000      -0.075      -0.040
urban             -0.2190      0.045     -4.814      0.000      -0.308      -0.130
electric          -0.3207      0.063     -5.076      0.000      -0.445      -0.197
bicycle            0.2046      0.048      4.279      0.000       0.111       0.298
nevermarr         -2.2501      0.202    -11.158      0.000      -2.645      -1.855
idlnchld_noans     0.6565      0.216      3.043      0.002       0.234       1.079
heduc_noans       -0.8853      0.191     -4.638      0.000      -1.259      -0.511
usemeth_noans      0.7732      0.212      3.639      0.000       0.357       1.190
==============================================================================
Omnibus:                      224.096   Durbin-Watson:                   1.886
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              856.760
Skew:                           0.004   Prob(JB):                    9.06e-187
Kurtosis:                       5.175   Cond. No.                         345.
==============================================================================

Warnings:
[1] Standard Errors are heteroscedasticity robust (HC1)
"""

* С увеличением возраста женщины на 1 год среднее количество детей возрастает на `0.17` (`p<0.001`, 95% доверительный интервал — `[0.16, 0.18]`)
* У женщин, не знающих, какое количество детей идеально, в среднем на `0.58` детей больше
* У женщин, никогда не выходивших замуж, при прочих равных в среднем на `2.3` ребёнка меньше (`p<0.001`, 95% доверительный интервал для разницы между средними — `[-2.6, -1.9]`)